File: typescript/src/bittorrent/extensions.ts
// SPDX-License-Identifier: MIT
// BEP-10 extension protocol + BEP-9 ut_metadata + BEP-11 ut_pex.
import { createHash } from 'node:crypto';
import { decode, decodeN, encode, type BValue } from './bencode.js';
import { decodeCompactPeers, encodeCompactPeers, type PeerAddr } from './dht.js';

export const EXTENDED_MESSAGE_ID = 20;
export const EXTENSION_HANDSHAKE_ID = 0;
export const METADATA_REQUEST = 0;
export const METADATA_DATA = 1;
export const METADATA_REJECT = 2;
export const METADATA_PIECE_SIZE = 16384;

export function wrapExtended(subId: number, body: Buffer): Buffer {
  return Buffer.concat([Buffer.from([subId]), body]);
}

export function splitExtended(payload: Buffer): [number, Buffer] {
  if (payload.length < 1) throw new Error('empty extended payload');
  return [payload[0], Buffer.from(payload.subarray(1))];
}

export function buildExtensionHandshake(supported: { [k: string]: number }, metadataSize = 0): Buffer {
  const m: { [k: string]: BValue } = {};
  for (const [k, v] of Object.entries(supported)) m[k] = v;
  const d: { [k: string]: BValue } = { m };
  if (metadataSize > 0) d['metadata_size'] = metadataSize;
  return wrapExtended(EXTENSION_HANDSHAKE_ID, encode(d));
}

export function parseExtensionHandshake(body: Buffer): { supported: { [k: string]: number }; metadataSize: number } {
  const d = decode(body) as { [k: string]: BValue };
  const supported: { [k: string]: number } = {};
  const m = (d['m'] ?? {}) as { [k: string]: BValue };
  for (const [k, v] of Object.entries(m)) supported[k] = v as number;
  return { supported, metadataSize: (d['metadata_size'] as number) ?? 0 };
}

export function buildMetadataRequest(piece: number): Buffer {
  return encode({ msg_type: METADATA_REQUEST, piece });
}

export function buildMetadataData(piece: number, totalSize: number, data: Buffer): Buffer {
  return Buffer.concat([encode({ msg_type: METADATA_DATA, piece, total_size: totalSize }), data]);
}

export function buildMetadataReject(piece: number): Buffer {
  return encode({ msg_type: METADATA_REJECT, piece });
}

export function parseMetadata(body: Buffer): { type: number; piece: number; totalSize: number; data: Buffer } {
  const [v, n] = decodeN(body, 0);
  const d = v as { [k: string]: BValue };
  return {
    type: d['msg_type'] as number,
    piece: d['piece'] as number,
    totalSize: (d['total_size'] as number) ?? 0,
    data: Buffer.from(body.subarray(n)),
  };
}

export class MetadataAssembler {
  private pieces = new Map<number, Buffer>();
  constructor(public totalSize: number) {}

  pieceCount(): number {
    return Math.ceil(this.totalSize / METADATA_PIECE_SIZE);
  }

  add(piece: number, data: Buffer): void {
    this.pieces.set(piece, Buffer.from(data));
  }

  isComplete(): boolean {
    return this.pieces.size === this.pieceCount();
  }

  tryFinish(infoHash: Buffer): Buffer | null {
    if (!this.isComplete()) return null;
    const parts: Buffer[] = [];
    for (let i = 0; i < this.pieceCount(); i++) parts.push(this.pieces.get(i) as Buffer);
    const out = Buffer.concat(parts);
    if (out.length !== this.totalSize) return null;
    if (!createHash('sha1').update(out).digest().equals(infoHash)) return null;
    return out;
  }
}

export function buildPexAdded(added: PeerAddr[]): Buffer {
  return encode({ added: encodeCompactPeers(added) });
}

export function parsePexAdded(body: Buffer): PeerAddr[] {
  const d = decode(body) as { [k: string]: BValue };
  if ('added' in d) return decodeCompactPeers(d['added'] as Buffer);
  return [];
}

File: go/bittorrent/extensions.go
// SPDX-License-Identifier: MIT

package bittorrent

import (
	"crypto/sha1"
	"fmt"
)

// ── BEP-10 extension protocol ────────────────────────────────────────────────

// ExtendedMessageID is the peer-wire message id for extended messages (BEP-10).
const ExtendedMessageID = 20

// ExtensionHandshakeID is the extended sub-message id of the handshake.
const ExtensionHandshakeID = 0

// WrapExtended builds an extended message payload: [subID][body]. This is the payload
// of a peer-wire Extended (id 20) message.
func WrapExtended(subID byte, body []byte) []byte {
	out := make([]byte, 1+len(body))
	out[0] = subID
	copy(out[1:], body)
	return out
}

// SplitExtended splits an extended payload into its sub-message id and body.
func SplitExtended(payload []byte) (byte, []byte, error) {
	if len(payload) < 1 {
		return 0, nil, fmt.Errorf("empty extended payload")
	}
	return payload[0], payload[1:], nil
}

// BuildExtensionHandshake builds a BEP-10 handshake advertising supported extensions
// (name → local sub-message id) and optionally the metadata size.
func BuildExtensionHandshake(supported map[string]int, metadataSize int) []byte {
	m := NewBDict()
	for name, id := range supported {
		_ = m.Add(name, BInt(int64(id)))
	}
	d := NewBDict()
	_ = d.Add("m", m)
	if metadataSize > 0 {
		_ = d.Add("metadata_size", BInt(int64(metadataSize)))
	}
	return WrapExtended(ExtensionHandshakeID, Encode(d))
}

// ExtensionHandshake is a parsed BEP-10 handshake.
type ExtensionHandshake struct {
	Supported    map[string]int
	MetadataSize int
}

// MetadataMessageID is the peer's ut_metadata sub-message id, or 0 if unsupported.
func (h ExtensionHandshake) MetadataMessageID() int { return h.Supported["ut_metadata"] }

// PexMessageID is the peer's ut_pex sub-message id, or 0 if unsupported.
func (h ExtensionHandshake) PexMessageID() int { return h.Supported["ut_pex"] }

// ParseExtensionHandshake parses a BEP-10 handshake body (the bencode dict after the sub-id).
func ParseExtensionHandshake(body []byte) (ExtensionHandshake, error) {
	h := ExtensionHandshake{Supported: map[string]int{}}
	v, err := Decode(body)
	if err != nil {
		return h, err
	}
	d, err := AsDict(v)
	if err != nil {
		return h, err
	}
	if mVal, ok := d.Get("m"); ok {
		if md, err := AsDict(mVal); err == nil {
			for _, name := range md.Keys() {
				if idVal, ok := md.Get(name); ok {
					if id, err := AsInt(idVal); err == nil {
						h.Supported[name] = int(id)
					}
				}
			}
		}
	}
	if sizeVal, ok := d.Get("metadata_size"); ok {
		if n, err := AsInt(sizeVal); err == nil {
			h.MetadataSize = int(n)
		}
	}
	return h, nil
}

// ── BEP-9 ut_metadata ────────────────────────────────────────────────────────

// MetadataMessageType is a ut_metadata message type.
type MetadataMessageType int

const (
	MetadataRequest MetadataMessageType = 0
	MetadataData    MetadataMessageType = 1
	MetadataReject  MetadataMessageType = 2
)

// MetadataPieceSize is the ut_metadata piece size (16 KiB).
const MetadataPieceSize = 16384

// BuildMetadataRequest builds a ut_metadata request for a piece.
func BuildMetadataRequest(piece int) []byte {
	d := NewBDict()
	_ = d.Add("msg_type", BInt(int64(MetadataRequest)))
	_ = d.Add("piece", BInt(int64(piece)))
	return Encode(d)
}

// BuildMetadataData builds a ut_metadata data message (bencode header + raw piece bytes).
func BuildMetadataData(piece, totalSize int, data []byte) []byte {
	d := NewBDict()
	_ = d.Add("msg_type", BInt(int64(MetadataData)))
	_ = d.Add("piece", BInt(int64(piece)))
	_ = d.Add("total_size", BInt(int64(totalSize)))
	return append(Encode(d), data...)
}

// BuildMetadataReject builds a ut_metadata reject message.
func BuildMetadataReject(piece int) []byte {
	d := NewBDict()
	_ = d.Add("msg_type", BInt(int64(MetadataReject)))
	_ = d.Add("piece", BInt(int64(piece)))
	return Encode(d)
}

// MetadataMessage is a parsed ut_metadata message.
type MetadataMessage struct {
	Type      MetadataMessageType
	Piece     int
	TotalSize int
	Data      []byte
}

// ParseMetadata parses a ut_metadata message, splitting the trailing raw piece bytes
// from the leading bencode dict.
func ParseMetadata(body []byte) (MetadataMessage, error) {
	var m MetadataMessage
	v, n, err := DecodeN(body)
	if err != nil {
		return m, err
	}
	d, err := AsDict(v)
	if err != nil {
		return m, err
	}
	if t, ok := d.Get("msg_type"); ok {
		ti, _ := AsInt(t)
		m.Type = MetadataMessageType(ti)
	}
	if p, ok := d.Get("piece"); ok {
		pi, _ := AsInt(p)
		m.Piece = int(pi)
	}
	if ts, ok := d.Get("total_size"); ok {
		tsi, _ := AsInt(ts)
		m.TotalSize = int(tsi)
	}
	m.Data = append([]byte(nil), body[n:]...)
	return m, nil
}

// MetadataAssembler reassembles the info dictionary from ut_metadata pieces and verifies
// it against the expected info-hash.
type MetadataAssembler struct {
	totalSize int
	pieces    map[int][]byte
}

// NewMetadataAssembler creates an assembler for a metadata of totalSize bytes.
func NewMetadataAssembler(totalSize int) *MetadataAssembler {
	return &MetadataAssembler{totalSize: totalSize, pieces: map[int][]byte{}}
}

// PieceCount is the number of 16 KiB pieces.
func (a *MetadataAssembler) PieceCount() int {
	return (a.totalSize + MetadataPieceSize - 1) / MetadataPieceSize
}

// Add stores a metadata piece.
func (a *MetadataAssembler) Add(piece int, data []byte) {
	a.pieces[piece] = append([]byte(nil), data...)
}

// IsComplete reports whether every piece is present.
func (a *MetadataAssembler) IsComplete() bool { return len(a.pieces) == a.PieceCount() }

// TryFinish assembles the info dict and returns it if it matches infoHash.
func (a *MetadataAssembler) TryFinish(infoHash [20]byte) ([]byte, bool) {
	if !a.IsComplete() {
		return nil, false
	}
	out := make([]byte, 0, a.totalSize)
	for i := 0; i < a.PieceCount(); i++ {
		out = append(out, a.pieces[i]...)
	}
	if len(out) != a.totalSize {
		return nil, false
	}
	if sha1.Sum(out) != infoHash {
		return nil, false
	}
	return out, true
}

// ── BEP-11 ut_pex ────────────────────────────────────────────────────────────

// BuildPexAdded builds a ut_pex message advertising added peers (compact form).
func BuildPexAdded(added []PeerAddr) []byte {
	d := NewBDict()
	_ = d.Add("added", BStr(EncodeCompactPeers(added)))
	return Encode(d)
}

// ParsePexAdded parses the "added" peers from a ut_pex message.
func ParsePexAdded(body []byte) ([]PeerAddr, error) {
	v, err := Decode(body)
	if err != nil {
		return nil, err
	}
	d, err := AsDict(v)
	if err != nil {
		return nil, err
	}
	if a, ok := d.Get("added"); ok {
		b, err := AsBytes(a)
		if err != nil {
			return nil, err
		}
		return DecodeCompactPeers(b)
	}
	return nil, nil
}
